src/OpenTelemetry.Instrumentation.Dependencies/Implementation/SqlEventSourceListener.netfx.cs

```diff
@@ -0,0 +1,131 @@
+
+        private static readonly Version Version = typeof(SqlEventSourceListener).Assembly.GetName().Version;
+        private static readonly ActivitySource SqlClientActivitySource = new ActivitySource(ActivitySourceName, Version.ToString());
+        private static readonly EventSource SqlEventSource = (EventSource)typeof(SqlConnection).Assembly.GetType("System.Data.SqlEventSource")?.GetField("Log", BindingFlags.Static | BindingFlags.NonPublic)?.GetValue(null);
+
+        private readonly SqlClientInstrumentationOptions options;
+
+        public SqlEventSourceListener(SqlClientInstrumentationOptions options = null)
+        {
+            this.options = options ?? new SqlClientInstrumentationOptions();
+
+            if (SqlEventSource != null)
+            {
+                this.EnableEvents(SqlEventSource, EventLevel.Informational);
+            }
+        }
+
+        public override void Dispose()
+        {
+            if (SqlEventSource != null)
+            {
+                this.DisableEvents(SqlEventSource);
+            }
+
+            base.Dispose();
+        }
+
+        protected override void OnEventWritten(EventWrittenEventArgs eventData)
+        {
+            if (eventData.EventId == 1)
+            {
+                // BeginExecuteEventId
+
+                var activity = SqlClientActivitySource.StartActivity(ActivityName, ActivityKind.Client);
+                if (activity == null)
+                {
+                    return;
+                }
+
+                activity.DisplayName = (string)eventData.Payload[2];
+
+                if (activity.IsAllDataRequested)
+                {
+                    activity.AddTag(SpanAttributeConstants.ComponentKey, "sql");
+
+                    activity.AddTag(SpanAttributeConstants.DatabaseTypeKey, "sql");
+                    activity.AddTag(SpanAttributeConstants.PeerServiceKey, (string)eventData.Payload[1]);
+                    activity.AddTag(SpanAttributeConstants.DatabaseInstanceKey, (string)eventData.Payload[2]);
+
+                    if (string.IsNullOrEmpty((string)eventData.Payload[3]))
+                    {
+                        activity.AddTag(SpanAttributeConstants.DatabaseStatementKey, "Text");
+                    }
+                    else
+                    {
+                        activity.AddTag(SpanAttributeConstants.DatabaseStatementKey, "StoredProcedure");
+                        if (this.options.CaptureStoredProcedureCommandName)
+                        {
+                            activity.AddTag(SpanAttributeConstants.DatabaseStatementKey, (string)eventData.Payload[3]);
+                        }
+                    }
+                }
+            }
+            else if (eventData.EventId == 2)
+            {
+                // EndExecuteEventId
+
+                var activity = Activity.Current;
+                if (activity == null || activity.Source != SqlClientActivitySource)
+                {
+                    return;
```

Comment:
we dont ever expect to this condition right? I'd suggest to add a Warning/Error log.

Reply:
Wouldn't this happen for samplers that return `ActivityDataRequest.None` which would result in no `Activity` being created? In that case Begin & End events would still fire, but `Activity` would be `null`.

Reply:
Yes you are right. This part is still not expected - ` activity.Source != SqlClientActivitySource` ?

Reply:
In my mind, it's expected. That case would be, sampler returned `ActivityDataRequest.None` for our Sql activity but there is _some_ other parent `Activity` running. It could be any `Activity`, really. Some parent that was sampled, and the child was `None`, or it could be some random `Activity` not run through a sampler at all. Something like this:

```csharp
            Activity myActivity = new Activity("Call SQL user activity");
            myActivity.Start();
            try
            {
                using (SqlConnection connection = new SqlConnection("Data Source=sqlserver;Initial Catalog=master;Integrated Security=True;"))
                {
                    connection.Open();

                    using (SqlCommand command = new SqlCommand("select 1/1", connection))
                    {
                        command.ExecuteNonQuery();
                    }
                }
            }
            catch { }
            myActivity.Stop();
            Log.LogActivity(myActivity);
```

In that case, when `EndExecute` fires `Activity.Current` will be non-null! The `activity.Source == SqlClientActivitySource` check is to make sure it is actually _our_ activity so we don't stop/tag something that doesn't belong to us.

Reply:
Perfect! thanks for clarifying.